### PR TITLE
update, get global isolation-groups handler endpoints & mapping

### PR DESCRIPTION
**Pull request by davidporter-id-au**
**What changed?**
- Adds endpoints for Update, Get global isolation-groups
- Adds E2E unit-tests which give some confidence that the library and handlers are working in integration, without being actual integration-tests. 

**Why?**

**How did you test it?**

**Potential risks**

**Release notes**

**Documentation Changes**

### CODE REVIEW

File: service/frontend/adminHandler.go [hunk 1]
```diff
@@ -1727,11 +1727,32 @@ func (adh *adminHandlerImpl) ListDynamicConfig(ctx context.Context, request *typ
...
+	defer func() { log.CapturePanic(recover(), adh.GetLogger(), &retError) }()
+	scope, sw := adh.startRequestProfile(ctx, metrics.GetGlobalIsolationGroups)
+	defer sw.Stop()
+
+	if request == nil {
+		return nil, adh.error(errRequestNotSet, scope)
+	}
+	ig := adh.Resource.GetIsolationGroupState()
+	if ig == nil {
+		return nil, nil
```
Comment (Shaddoll) [inline service/frontend/adminHandler.go:1739]: I think it's better to return `&types.GetGlobalIsolationGroupsResponse{}, nil`

Comment (davidporter-id-au) [reply service/frontend/adminHandler.go:1739]: good call

File: service/frontend/adminHandler.go [hunk 2]
```diff
@@ -1727,11 +1727,32 @@ func (adh *adminHandlerImpl) ListDynamicConfig(ctx context.Context, request *typ
...
 func (adh *adminHandlerImpl) GetGlobalIsolationGroups(ctx context.Context, request *types.GetGlobalIsolationGroupsRequest) (_ *types.GetGlobalIsolationGroupsResponse, retError error) {
-	panic("not implemented")
+	defer func() { log.CapturePanic(recover(), adh.GetLogger(), &retError) }()
+	scope, sw := adh.startRequestProfile(ctx, metrics.GetGlobalIsolationGroups)
+	defer sw.Stop()
+
+	if request == nil {
+		return nil, adh.error(errRequestNotSet, scope)
+	}
+	ig := adh.Resource.GetIsolationGroupState()
```
[the remark this answers is not in the document]
Comment (davidporter-id-au) [reply service/frontend/adminHandler.go:1737]: Not sure I understand? Where ought I get this from?

File: common/types/admin.go
Comment (Shaddoll) [inline common/types/admin.go:1]: missing tests
you can check new missing lines in coverall

File: common/metrics/defs.go [hunk 3]
```diff
@@ -545,6 +545,10 @@ const (
 	AdminClientRestoreDynamicConfigScope
 	// AdminClientListDynamicConfigScope tracks RPC calls to admin service
 	AdminClientListDynamicConfigScope
+	// AdminClientGetGlobalIsolationGroupsScope is a request to get all the global isolation-groups
```
Comment (Shaddoll) [inline common/metrics/defs.go:548]: the change doesn't belong to this PR

Comment (davidporter-id-au) [reply common/metrics/defs.go:548]: whoops, sorry

File: common/types/admin.go
Comment (davidporter-id-au) [reply common/types/admin.go:1]: yes, I missed them as they were from the Admin CLi update and I'd accidentally carried them across. that was a bit sloppy of me, I'll try and not do that again.

File: common/types/mapper/proto/admin.go [hunk 4]
```diff
@@ -1165,21 +1167,53 @@ func ToDynamicConfigFilter(t *adminv1.DynamicConfigFilter) *types.DynamicConfigF
...
 
+func FromGetGlobalIsolationGroupsRequest(t *types.GetGlobalIsolationGroupsRequest) *adminv1.GetGlobalIsolationGroupsRequest {
+	if t == nil {
+		return nil
+	}
+	return &adminv1.GetGlobalIsolationGroupsRequest{}
+}
+
+func FromUpdateGlobalIsolationGroupsRequest(t *types.UpdateGlobalIsolationGroupsRequest) *adminv1.UpdateGlobalIsolationGroupsRequest {
+	if t == nil || len(t.IsolationGroups) == 0 {
```
Comment (Shaddoll) [inline common/types/mapper/proto/admin.go:1189]: nil and the non-nil should be treated differently
this is just for the purpose of keeping the behavior consistent
if you search `t == nil` in this file, they are all converted to `nil`

File: common/types/mapper/proto/admin.go [hunk 5]
```diff
@@ -1206,7 +1240,7 @@ func FromUpdateGlobalIsolationGroupsResponse(t *types.UpdateGlobalIsolationGroup
 
 func ToUpdateGlobalIsolationGroupsRequest(t *adminv1.UpdateGlobalIsolationGroupsRequest) *types.UpdateGlobalIsolationGroupsRequest {
 	if t == nil {
-		return nil
+		return &types.UpdateGlobalIsolationGroupsRequest{}
```
Comment (Shaddoll) [inline common/types/mapper/proto/admin.go:1243]: same as above

File: common/types/mapper/thrift/admin.go [hunk 6]
```diff
@@ -727,19 +755,35 @@ func FromUpdateGlobalIsolationGroupsResponse(t *types.UpdateGlobalIsolationGroup
...
+	if t == nil || t.IsolationGroups == nil {
+		return &admin.UpdateGlobalIsolationGroupsRequest{}
+	}
+	return &admin.UpdateGlobalIsolationGroupsRequest{
+		IsolationGroups: isolationGroupConfigToIDL(&t.IsolationGroups),
+	}
+}
+
 func ToUpdateGlobalIsolationGroupsRequest(t *admin.UpdateGlobalIsolationGroupsRequest) *types.UpdateGlobalIsolationGroupsRequest {
 	if t == nil {
```
Comment (Shaddoll) [inline common/types/mapper/thrift/admin.go:768]: same as above

File: common/types/mapper/proto/admin_test.go
Comment (Shaddoll) [inline common/types/mapper/proto/admin_test.go:1]: There doesn't seem to be any tests for `GetRequest` and `UpdateResponse`

File: service/frontend/adminHandler.go [hunk 7]
```diff
@@ -1727,11 +1727,32 @@ func (adh *adminHandlerImpl) ListDynamicConfig(ctx context.Context, request *typ
...
 
 func (adh *adminHandlerImpl) UpdateGlobalIsolationGroups(ctx context.Context, request *types.UpdateGlobalIsolationGroupsRequest) (_ *types.UpdateGlobalIsolationGroupsResponse, retError error) {
-	panic("Not implemented")
+	defer func() { log.CapturePanic(recover(), adh.GetLogger(), &retError) }()
+	scope, sw := adh.startRequestProfile(ctx, metrics.UpdateGlobalIsolationGroups)
+	defer sw.Stop()
+	if request == nil {
+		return nil, adh.error(errRequestNotSet, scope)
+	}
+	err := adh.GetIsolationGroupState().UpdateGlobalState(ctx, *request)
```
Comment (Shaddoll) [inline service/frontend/adminHandler.go:1751]: `adh.GetIsolationGroupState()` requires a `nil` check

File: service/frontend/adminHandler.go [hunk 8]
```diff
@@ -1727,11 +1727,32 @@ func (adh *adminHandlerImpl) ListDynamicConfig(ctx context.Context, request *typ
...
+	defer func() { log.CapturePanic(recover(), adh.GetLogger(), &retError) }()
+	scope, sw := adh.startRequestProfile(ctx, metrics.GetGlobalIsolationGroups)
+	defer sw.Stop()
+
+	if request == nil {
+		return nil, adh.error(errRequestNotSet, scope)
+	}
+	ig := adh.GetIsolationGroupState()
+	if ig == nil {
+		return &types.GetGlobalIsolationGroupsResponse{}, nil
```
Comment (Shaddoll) [inline service/frontend/adminHandler.go:1739]: should we return an error telling the client that isolation group state is not enabled?

File: service/frontend/adminHandler_test.go [hunk 9]
```diff
@@ -764,3 +776,207 @@ func (s *adminHandlerSuite) Test_GetDynamicConfig_FilterMatch() {
 	s.NoError(err)
 	s.Equal(resp.Value.Data, encTrue)
 }
+
+func Test_GetGlobalIsolationGroups(t *testing.T) {
```
Comment (Shaddoll) [inline service/frontend/adminHandler_test.go:780]: the test is more like an integration test
I think we should mock `IsolationGroupState` instead of mocking its dependencies.

File: common/types/admin.go [hunk 10]
```diff
@@ -392,6 +394,18 @@ type IsolationGroupPartition struct {
 // Indicating that task processing isn't to occur within this isolationGroup anymore, but all others are ok.
 type IsolationGroupConfiguration map[string]IsolationGroupPartition
 
+func (i IsolationGroupConfiguration) ToPartitionList() []IsolationGroupPartition {
```
Comment (Shaddoll) [inline common/types/admin.go:397]: a simple test can be added for this func

File: common/types/mapper/proto/admin_test.go
Comment (davidporter-id-au) [reply common/types/mapper/proto/admin_test.go:1]: This is true, they're empty requests so I didn't add them, I want to focus coverage on areas such as the partition library which makes me more worried.

File: common/types/admin.go (hunk 10 above)
Comment (davidporter-id-au) [reply common/types/admin.go:397]: I can, however this is covered in the test `TestIsDrainedHandler` so I'm unclear why it's not appearing as covered in the coverage...

File: common/types/mapper/thrift/admin.go [hunk 11]
```diff
@@ -727,19 +755,35 @@ func FromUpdateGlobalIsolationGroupsResponse(t *types.UpdateGlobalIsolationGroup
 	return &admin.UpdateGlobalIsolationGroupsResponse{}
 }
 
+func FromUpdateGlobalIsolationGroupsRequest(t *types.UpdateGlobalIsolationGroupsRequest) *admin.UpdateGlobalIsolationGroupsRequest {
+	if t == nil || t.IsolationGroups == nil {
+		return &admin.UpdateGlobalIsolationGroupsRequest{}
```
[the remark this answers is not in the document]
Comment (davidporter-id-au) [reply common/types/mapper/thrift/admin.go:763]: This is slightly different, it's a subfield which is nil, and in that I'm returning as nil

File: service/frontend/adminHandler.go (hunk 8 above)
Comment (davidporter-id-au) [reply service/frontend/adminHandler.go:1742]: Ah sorry I didn't read correctly, your comment, yes, done so

File: service/frontend/adminHandler.go [hunk 12]
```diff
@@ -1727,11 +1727,38 @@ func (adh *adminHandlerImpl) ListDynamicConfig(ctx context.Context, request *typ
...
 func (adh *adminHandlerImpl) UpdateGlobalIsolationGroups(ctx context.Context, request *types.UpdateGlobalIsolationGroupsRequest) (_ *types.UpdateGlobalIsolationGroupsResponse, retError error) {
-	panic("Not implemented")
+	defer func() { log.CapturePanic(recover(), adh.GetLogger(), &retError) }()
+	scope, sw := adh.startRequestProfile(ctx, metrics.UpdateGlobalIsolationGroups)
+	defer sw.Stop()
+	if request == nil {
+		return nil, adh.error(errRequestNotSet, scope)
+	}
+	if adh.GetIsolationGroupState() == nil {
+		adh.error(types.BadRequestError{Message: "isolation groups are not enabled in this cluster"}, scope)
```
Comment (Shaddoll) [inline service/frontend/adminHandler.go:1755]: nit: missing `return`

File: service/frontend/adminHandler.go [hunk 13]
```diff
@@ -1727,11 +1727,38 @@ func (adh *adminHandlerImpl) ListDynamicConfig(ctx context.Context, request *typ
...
-	panic("not implemented")
+	defer func() { log.CapturePanic(recover(), adh.GetLogger(), &retError) }()
+	scope, sw := adh.startRequestProfile(ctx, metrics.GetGlobalIsolationGroups)
+	defer sw.Stop()
+
+	if request == nil {
+		return nil, adh.error(errRequestNotSet, scope)
+	}
+	if adh.GetIsolationGroupState() == nil {
+		adh.error(types.BadRequestError{Message: "isolation groups are not enabled in this cluster"}, scope)
```
Comment (Shaddoll) [inline service/frontend/adminHandler.go:1738]: missing `return` and the following check is duplicate

File: service/frontend/adminHandler.go (hunk 12 above)
Comment (davidporter-id-au) [reply service/frontend/adminHandler.go:1755]: 🤦

**Review by Shaddoll — APPROVED** [review]
